dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -514,79 +582,99 @@ bool FramesDecoder::ReadRegularFrame(uint8_t *data, bool copy_to_output) {
-  int keyframe_id = frame_entry.last_keyframe_id;
-  auto &keyframe_entry = Index(keyframe_id);
+  if (frame_id == next_frame_idx_) {
+    return;  // No need to seek
+  }
 
-  LOG_LINE << "Seeking to frame " << frame_id << " timestamp " << frame_entry.pts << std::endl;
+  // If we are seeking to a frame that is before the current frame,
+  // or we are seeking to a frame that is more than 10 frames away,
+  // we will to seek to the nearest keyframe first
+  LOG_LINE << "SeekFrame: frame_id=" << frame_id << ", next_frame_idx_=" << next_frame_idx_
+           << std::endl;
+  if (frame_id < next_frame_idx_ || frame_id > next_frame_idx_ + 10) {
+    // If we have an index, we can seek to the nearest keyframe first
+    if (HasIndex()) {
+      LOG_LINE << "Using index to find nearest keyframe" << std::endl;
+      const auto &current_frame = Index(next_frame_idx_);
+      const auto &requested_frame = Index(frame_id);
+      auto keyframe_id = requested_frame.last_keyframe_id;
+      // if we are seeking to a different keyframe than the current frame,
+      // or if we are seeking to a frame that is before the current frame,
+      // we need to seek to the keyframe first
+      if (current_frame.last_keyframe_id != keyframe_id || frame_id < next_frame_idx_) {
+        // We are seeking to a different keyframe than the current frame,
+        // so we need to seek to the keyframe first
+        auto &keyframe_entry = Index(keyframe_id);
+        LOG_LINE << "Seeking to key frame " << keyframe_id << " timestamp " << keyframe_entry.pts
+                 << " for requested frame " << frame_id << " timestamp " << requested_frame.pts
+                 << std::endl;
+
+        // Seeking clears av buffers, so reset flush state info
+        if (flush_state_) {
+          while (ReadFlushFrame(nullptr, false)) {}
+          flush_state_ = false;
+        }
 
-  // Seeking clears av buffers, so reset flush state info
-  if (flush_state_) {
-    while (ReadFlushFrame(nullptr, false)) {}
-    flush_state_ = false;
+        int ret = av_seek_frame(av_state_->ctx_, av_state_->stream_id_, keyframe_entry.pts,
+                                AVSEEK_FLAG_FRAME);
+        DALI_ENFORCE(ret >= 0, make_string("Failed to seek to keyframe", keyframe_id, "in video \"",
+                                           Filename(), "\" due to ", detail::av_error_string(ret)));
+        avcodec_flush_buffers(av_state_->codec_ctx_);
+        next_frame_idx_ = keyframe_id;
+      }
+    } else if (frame_id < next_frame_idx_) {
+      LOG_LINE << "No index available and seeking backwards, resetting decoder" << std::endl;
+      // If we are seeking to a frame that is before the current frame and there's no index,
+      // we need to reset the decoder
+      if (IsFormatSeekable()) {
+        Reset();
+      } else {
+        memory_video_file_->Seek(0, SEEK_SET);
+        av_state_ = std::make_unique<AvState>();
+        next_frame_idx_ = 0;
+      }
```

Comment:
I just checked that. It doesn't work and crashes :(

Reply:
done